Port iseq GC callbacks

XrXr opened the pull request:
rb_yjit_iseq_free() is close to the allocation site for the payload
cause they are related.

XrXr committed:
Port iseq GC callbacks

rb_yjit_iseq_free() is close to the allocation site for the payload
cause they are related.

XrXr committed:
Add VALUE->{iseq,cme} downcast and move type assert there

I realized that during compaction the VALUE we store could be `T_MOVED`
before we pass it to `rb_gc_location`. Asserting in downcast makes
more sense.

XrXr commented:
A bit of a yak shave to add conversion methods between VALUE and `{cme,iseq}` but 🤷🏼 maybe it'll help with debugging.

noahgibbs commented on yjit/src/core.rs:
```diff
@@ -377,6 +378,158 @@ fn get_iseq_payload(iseq: IseqPtr) -> &'static mut IseqPayload
     unsafe { payload_non_null.as_mut() }.unwrap()
 }
 
+/// Free the per-iseq YJTI payload
```
Typo: YJTI -> YJIT

maximecb commented on yjit/src/core.rs:
```diff
@@ -377,6 +378,158 @@ fn get_iseq_payload(iseq: IseqPtr) -> &'static mut IseqPayload
 
+/// Free the per-iseq YJTI payload
+#[no_mangle]
+pub extern "C" fn rb_yjit_iseq_free(payload: *mut c_void)
+{
+    let payload = {
+        if payload.is_null() {
+            // Nothing to free.
+            return;
+        } else {
+            payload as *mut IseqPayload
+        }
+    };
```
Detail maybe, but you could pass an IseqPtr to these functions instead. Then you'd have access to the `get_iseq_payload` method which does the unwrapping, and you'd be able to print the iseq pointer for debugging (if ever needed).

XrXr replied:
`get_iseq_payload` allocates the payload when it's null though so to make that change I would need to touch some other functions -- wouldn't want to be allocating and then immediately freeing the payload. I think it's a good call to make the change for other reasons but it feels out of scope for this PR.

maximecb reviewed (commented):
Looks good with minor comments 👍

XrXr commented:
CI caught an issue with RefCell borrowing so I'm looking into that. (thanks for adding the temporary CI check 👍🏼 )

```
thread '<unnamed>' panicked at 'already mutably borrowed: BorrowError', src/core.rs:430:37
```

maximecb commented:
> CI caught an issue with RefCell borrowing so I'm looking into that. (thanks for adding the temporary CI check 👍🏼 )
> 
> ```
> thread '<unnamed>' panicked at 'already mutably borrowed: BorrowError', src/core.rs:430:37
> ```

I noticed that error in other failing tests too 👍 Thanks for letting me know you'll be looking into it :)

XrXr committed:
Don't hold mutable RefCell<Branch> borrow while compiling

When GC triggers while compiling we need to mark the branch
and we need to borrow. Mutable borrows prevents any further
borrows.
